truffle/src/main/java/org/jruby/truffle/RubyFiletypeDetector.java

```diff
@@ -0,0 +1,50 @@
+package org.jruby.truffle;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.spi.FileTypeDetector;
+import java.util.Locale;
+import java.util.regex.Pattern;
+
+public final class RubyFiletypeDetector extends FileTypeDetector {
+
+    private static final String[] KNOWN_RUBY_FILES = new String[]{ "Gemfile", "Rakefile", "Mavenfile" };
+    private static final String[] KNOWN_RUBY_SUFFIXES = new String[]{ ".rb", ".rake", ".gemspec" };
+    private static final String RUBY_MIME = "application/x-ruby";
+    private static final Pattern SHEBANG_REGEXP = Pattern.compile("^#! ?/usr/bin/(env +ruby|ruby).*");
+
+    @Override
+    public String probeContentType(Path path) throws IOException {
+        final Path fileNamePath = path.getFileName();
+        if (fileNamePath == null) {
+            return null;
+        }
+
+        final String fileName = fileNamePath.toString();
+
+        for (String candidate : KNOWN_RUBY_SUFFIXES) {
+            if (fileName.toLowerCase(Locale.ROOT).endsWith(candidate)) {
```

Comment:
It's probably not a huge gain because most files will be '.rb', but maybe lift the `toLowerCase` call above the `for` loop to avoid repeated calculations.